python-package/xgboost/spark/data.py

```diff
@@ -0,0 +1,262 @@
+
+
+def prepare_train_val_data(data_iterator,
+                           has_weight,
+                           has_validation,
+                           has_fit_base_margin=False):
+    def gen_data_pdf():
+        for pdf in data_iterator:
+            yield pdf
+
+    return _process_data_iter(gen_data_pdf(),
+                              train=True,
+                              has_weight=has_weight,
+                              has_validation=has_validation,
+                              has_fit_base_margin=has_fit_base_margin,
+                              has_predict_base_margin=False)
+
+
+def prepare_predict_data(data_iterator, has_predict_base_margin):
+    return _process_data_iter(data_iterator,
+                              train=False,
+                              has_weight=False,
+                              has_validation=False,
+                              has_fit_base_margin=False,
+                              has_predict_base_margin=has_predict_base_margin)
+
+
+def _check_feature_dims(num_dims, expected_dims):
+    """
+    Check all feature vectors has the same dimension
+    """
+    if expected_dims is None:
+        return num_dims
+    if num_dims != expected_dims:
+        raise ValueError("Rows contain different feature dimensions: "
+                         "Expecting {}, got {}.".format(
+                             expected_dims, num_dims))
+    return expected_dims
+
+
+def _row_tuple_list_to_feature_matrix_y_w(data_iterator, train, has_weight,
+                                          has_fit_base_margin,
+                                          has_predict_base_margin,
+                                          has_validation: bool = False):
+    """
+    Construct a feature matrix in ndarray format, label array y and weight array w
+    from the row_tuple_list.
+    If train == False, y and w will be None.
+    If has_weight == False, w will be None.
+    If has_base_margin == False, b_m will be None.
+    Note: the row_tuple_list will be cleared during
+    executing for reducing peak memory consumption
+    """
+    expected_feature_dims = None
+    label_list, weight_list, base_margin_list = [], [], []
+    label_val_list, weight_val_list, base_margin_val_list = [], [], []
+    values_list, values_val_list = [], []
+
+    # Process rows
+    for pdf in data_iterator:
+        if type(pdf) == tuple:
+            pdf = pd.concat(list(pdf), axis=1, names=["values", "baseMargin"])
+
+        if len(pdf) == 0:
+            continue
+        if train and has_validation:
+            pdf_val = pdf.loc[pdf["validationIndicator"], :]
+            pdf = pdf.loc[~pdf["validationIndicator"], :]
+
+        num_feature_dims = len(pdf["values"].values[0])
+
+        expected_feature_dims = _check_feature_dims(num_feature_dims,
+                                                    expected_feature_dims)
+
+        values_list.append(pdf["values"].to_list())
```

Comment:
Not an issue that needs to be fixed in the initial PR, but I think the `to_list` is not necessary and we can use some more optimized procedures in pandas or numpy.

Reply:
I marked a TODO there and we can address it in follow-up PR.

Reply:
addressed in this PR.